### PR TITLE
Implement delayed key event synthesis support for framework

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This implements the design in https://flutter.dev/go/handling-synchronous-keyboard-events for Android.

I started with Android, but this will be used for all platforms as we add them.

The related engine PR is: https://github.com/flutter/engine/pull/19024

## Related Issues

- https://github.com/flutter/flutter/issues/33521
- https://github.com/flutter/flutter/issues/47156

## Tests

- Added a test to make sure that generated key events send back messages indicating that they handled the event.

## Breaking Change

- [X] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by PixelToast** [comment]
What would happen if multiple Flutter Views are on top of each other?

I assume this would lead to n^2 key events in rapid succession, but its not clear if there would be any notable side effects of that.

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 1]
```diff
@@ -534,14 +574,28 @@ class RawKeyboard {
...
-    }
+    // Send the event to passive listeners.
     for (final ValueChanged<RawKeyEvent> listener in List<ValueChanged<RawKeyEvent>>.from(_listeners)) {
       if (_listeners.contains(listener)) {
         listener(event);
       }
     }
+    // Send the key event to the keyEventHandler, and send the appropriate
+    // response to the platform so that it can resolve the event's handling.
+    _sendKeyEventResponse(event, handled: keyEventHandler?.call(event) ?? false);
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:585]: Should we assert here that if the keyEventHandler was provided it returned true/false (and not null)? Assuming that returning null from keyEventHandler is illegal.

Comment (goderbauer) [reply packages/flutter/lib/src/services/raw_keyboard.dart:585]: If null is illegal, let's document that on RawKeyEventHandler as well.

File: packages/flutter/lib/src/services/raw_keyboard.dart [hunk 2]
```diff
@@ -494,20 +504,50 @@ class RawKeyboard {
...
+  /// [FocusManager]) to receive key events. If you are not using the
+  /// [FocusManager] to manage focus, then to be able to stop propagation of the
+  /// event, and indicate that the event was handled, set this attribute to a
+  /// [RawKeyEventHandler]. Otherwise, key events will be assumed to not have
+  /// been handled, and will also be sent to other (possibly non-Flutter)
+  /// controls in the application.
+  ///
+  /// See also:
+  ///
+  ///  * [Focus.onKey], a [Focus] callback attribute that will be given key
```
Comment (goderbauer) [inline packages/flutter/lib/src/services/raw_keyboard.dart:546]: maybe link to [addListener] as well from here and say how that's different then receiving the key events from here.

**Review by goderbauer — APPROVED** [review]
LGTM, just some doc nits.

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Comment by flutter-github-sync** [comment]
Started Google testing for this PR

**Comment by flutter-github-sync** [comment]
Google testing passed!

File: packages/flutter/lib/src/services/raw_keyboard.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/services/raw_keyboard.dart]: Done.

**Pull request merged**